Adds new maintenance spawn

cacogen opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

https://user-images.githubusercontent.com/25089914/155257039-d27eda80-d1b0-459c-a29d-67cba9c92f79.mp4

https://user-images.githubusercontent.com/25089914/155257195-2a22abda-0a9c-4f39-be49-54251530d504.mp4

Adds a system similar to smuggler's satchels that spawns five pairs of quantum pads hidden underneath the floor in both maintenance and secluded public areas. 

The pads try to be at least 50 tiles apart from each other to maximise their usefulness.

Unlike smuggler's satchels, they do not show up to observers or on t-ray scanners. Instead, they have a couple of quirks in the way they spawn that make them fairly easy to find. You can always brute force it, too.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

It's fun. It makes maintenance less linear. It makes it easier to lose pursuers in maintenance.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
add: Adds new maintenance spawn
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Time-Green commented:
I love this so much, but isn’t it a bit common? I think it’d be more fitting as a roundstart trait so it doesn’t get old

MrDoomBringer commented:
If it makes maintenance objectively more interesting and needs to be actively searched for, I don't see any need to make it any less common than ~3 pairs per round.

Also @cacogen I suggest updating your changelog, as I don't think players will understand what you mean by 'maintenance structure spawner', even if you're intentionally being vague.

tralezab commented on code/game/objects/structures/ladders.dm:
```diff
@@ -197,3 +197,61 @@
+	AddElement(/datum/element/undertile, TRAIT_T_RAY_VISIBLE, use_anchor = TRUE)
+	return ..()
+
+/obj/structure/ladder/unbreakable/teleport/LateInitialize()
```
this won't be called unless Initialize returns INITIALIZE_HINT_LATELOAD

cacogen replied:
The parent ladder does that (Line 25)

tralezab replied:
🙏

tralezab commented on code/game/objects/structures/ladders.dm:
```diff
@@ -197,3 +197,61 @@
+		if(target.overfloor_placed)
+			var/turf/open/floor/floorturf = target
+			if(floorturf.floor_tile)
+				var/obj/item/stack/tile/floor_tile = new floorturf.floor_tile(floorturf)
```
call spawn_tile() instead and set floor_tile as what it returns

tralezab commented on code/game/objects/structures/ladders.dm:
```diff
@@ -197,3 +197,61 @@
+/obj/structure/ladder/unbreakable/teleport/travel(going_up, mob/user, is_ghost, obj/structure/ladder/ladder)
+	var/turf/target = get_turf(ladder)
+
+	if(!is_ghost)
```
switch this around and early return after ghosts attempt this, you have a small else case at the end of this if proc that makes you have to indent everything

tralezab commented on code/game/objects/structures/ladders.dm:
```diff
@@ -197,3 +197,61 @@
+		user.zMove(target = target, z_move_flags = ZMOVE_CHECK_PULLEDBY|ZMOVE_ALLOW_BUCKLED|ZMOVE_INCLUDE_PULLED)
+
+		if(target.overfloor_placed)
+			var/turf/open/floor/floorturf = target
```
this is a pretty scary assumption that the destination's turf will be a floor. if it's not, this code will break when it tries spawning the floortile that nonfloors don't have.

tralezab reviewed (changes requested):
these are strictly the code-side stuff, i have some lame gamey things to comment on as well

tralezab commented:
ladders teleporting people across the map feels too gamey to me. any chance these could be rethemed as old abandoned warp pads, or whatever you can think up that makes sense?

D4C-420 commented:
I also second the idea of not using ladders because as is currently ladders are used strictly for moving between Z-levels. Not across which might confuse players upon first using them. Best suggestion I have would actually be having them be similar to jaunter portals where they teleport you but also make you throw up and stun you for a bit as a sort of cost of using them (also so you can't just rapidly jump in and out of them like ladders, combined with the fact that those portals from the mining tool already move you randomly  across the station Z level like these do working even on station.)

cacogen commented:
>I love this so much, but isn’t it a bit common? I think it’d be more fitting as a roundstart trait so it doesn’t get old

It needs to be secretive but practical. Like it needs to give you a good teleport distance (which is not something I've factored in yet) and they need to be hard to find without never being seen. The right amount is based on how difficult they are to find and how valuable they actually are.

The problems I have at the moment is that they always pick the same handful of spots which makes them very easy to metagame and they can teleport you a remarkably shit distance to the point where you feel ripped off and disappointed.

MrDoomBringer commented:
@cacogen I'm confused. Why not just generate the location of both pairs programmatically/randomly?

cacogen commented:
That was definitely the first thing I tried, but the results weren't very good. Iended up with ladders beside each other, or in the middle of hallways or inside of windows. 

I could fix the ladders spawning too close to each other or on turfs with dense contents but you'd still end up with ladders in the middle of maintenance corridors. By using spawners you can make sure they only appear in more out of the way places.

cacogen commented on code/game/machinery/quantum_pad.dm:
```diff
@@ -190,6 +191,23 @@
+
+/obj/machinery/quantumpad/maintenance/Initialize(mapload)
+	. = ..()
+	AddElement(/datum/element/undertile, TRAIT_T_RAY_VISIBLE, INVISIBILITY_OBSERVER)
```
This stopped working properly. The pads start above the floor and are only hidden when you remove and replace the tile. Don't know how to fix it.

cacogen commented:
@MrDoomBringer

It's now random again to prevent metagaming the spawners. But I've incorporated the code I made when I changed it to use spawners, meaning the pads will always be a good distance away from each other. 

I've also made sure they won't end up underneath things like windows. They'll still end up in places like the middle of a maintenance corridor but given the scarcity and the distance between pads it shouldn't be too much of an eyesore.

cacogen commented:
I've been waiting for a very long time for a review of this.

MrDoomBringer commented:
I've been waiting for a very long time for a review of this.

optimumtact commented:
probably because you went out of your way to make every maintainer actively dislike you and now you realise you aren't owed a review.

cacogen commented:
I’ve never thought I was owed a review. And I don’t think the lack of reviews has anything to do with how the maintainers may or may not feel about me.

cacogen committed:
Maintenance quantum pad spawner

YakumoChen commented:
I don't understand the reason T Ray doesn't detect this. Isn't that the whole point of them, to see stuff under floor tiles? It's inconsistent.

cacogen commented:
I tested this extensively and found them too easy to find. The easier they are to find the less of them you can have because they're too easy to metagame. This way you have to use a bit of deductive reasoning and trial and error to find them.

Ice Box wasn't a good example of a station to show this on though as the maintenance tunnels don't have a lot of floorspace and most of that isn't covered by tiles, so they're very easy to find. They're harder to find on DeltaStation.

There's a lore paper that spawns with them on a random empty tile in maintenance that explains they've been specifically designed to be undetectable by t-ray scanners.

cacogen commented:
the end of an era
